<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>How to use magenta • magenta</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="How to use magenta">
<meta property="og:description" content="magenta">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">magenta</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../articles/Introduction.html">Introduction</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Introduction.html">Introduction to magenta (Overview)</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/Loggers.html">Logger Definitions</a>
    </li>
    <li>
      <a href="../articles/Resistance.html">Resistance</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ojwatson/magenta">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Introduction_files/header-attrs-2.5/header-attrs.js"></script><link href="Introduction_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="Introduction_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>How to use magenta</h1>
                        <h4 class="author">OJ Watson</h4>
            
            <h4 class="date">2019-12-10</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/OJWatson/magenta/blob/master/vignettes/Introduction.Rmd"><code>vignettes/Introduction.Rmd</code></a></small>
      <div class="hidden name"><code>Introduction.Rmd</code></div>

    </div>

    
    
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p><code>**magenta**</code> is an individual-based simulation model of malaria epidemiology and parasite genetics.</p>
<p><code>magenta</code> extends the Imperial malaria model by tracking the infection history of individuals. With this additional genetic characteristics of the parasite can be assessed for looking at both neutral genetic variation as well as loci under selection.</p>
<p>The model is written in C++/Rcpp and interfaced with R.</p>
<hr>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>You can also install <code>rdhs</code> from github with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#install.packages("devtools")</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"OJwatson/rdhs"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load the package</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ojwatson.github.io/magenta/">magenta</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="demonstration" class="section level2">
<h2 class="hasAnchor">
<a href="#demonstration" class="anchor"></a>Demonstration</h2>
<div id="base-model" class="section level3">
<h3 class="hasAnchor">
<a href="#base-model" class="anchor"></a>Base Model</h3>
<p>The <code>magenta</code> simulation model is run using the <code>pipeline</code> function, which handles creating needed parameter objects, starting the simulation, summarising the simulation model state at desired intervals and handling how to save the final model output and state.</p>
<p>For example to run the model for 10 years at an EIR of 1 and no seasonality for a population of 10,000.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">out1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pipeline.html">pipeline</a></span><span class="op">(</span>EIR <span class="op">=</span> <span class="fl">1</span>, years <span class="op">=</span> <span class="fl">10</span>, N <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span>
<span class="co">## Seed set to 34491875</span>
<span class="co">## magenta v1.2.0</span></code></pre></div>
<p>By default, magenta will run the simulation for the entire duration and return a few summaries of the human population at the end of the 10 years, which are stored within the <code>Loggers</code> list in <code>out</code>. For example, the human proportions in each infection state are the last 6 elements in <code>Loggers</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">tail</a></span><span class="op">(</span><span class="va">out1</span><span class="op">$</span><span class="va">Loggers</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span>
<span class="co">##           S          D          A          U           T           P</span>
<span class="co">## 1 0.8608592 0.00156126 0.09441759 0.03743192 0.001026548 0.004703534</span></code></pre></div>
<p>(See <a href="https://ojwatson.github.io/magenta/articles/loggers.html">Loggers vignette</a> for more info about what is returned in Loggers).</p>
<p>If we want to save information about the model state over time we set <code>update_save</code> to <code>TRUE</code> and then specify at what intervals we want to stop the simulation using <code>update_length</code>. For example, to fetch information about the simualtion at monthly intervals:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">out2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pipeline.html">pipeline</a></span><span class="op">(</span>EIR <span class="op">=</span> <span class="fl">1</span>, years <span class="op">=</span> <span class="fl">10</span>, N <span class="op">=</span> <span class="fl">10000</span>, 
                 update_save <span class="op">=</span> <span class="cn">TRUE</span>, update_length <span class="op">=</span> <span class="fl">30</span>,
                 seed <span class="op">=</span> <span class="fl">123456789L</span><span class="op">)</span>
<span class="co">## Seed set to 123456789</span>
<span class="co">## magenta v1.2.0</span>
<span class="co">## Starting Stochastic Simulation for 10 years</span>
<span class="co">## Running: [=====================================================================================] 100% eta: 0s</span></code></pre></div>
<p>When doing <code>update_save</code> we now get a progress bar for the simulation and the returned object <code>out</code> is now a list where each the first n-1 elements are unnamed lists and are the saved information at each month. (The last element contains the pointer to the simulation, which allows us to carry on the simulation if needed.) For example to plot the size of the susceptible proportion over time:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">S</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">out2</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">out2</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span><span class="op">]</span>, <span class="st">"[["</span>, <span class="st">"S"</span><span class="op">)</span><span class="op">)</span>
<span class="va">time</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">out2</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">out2</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span><span class="op">]</span>, <span class="st">"[["</span>, <span class="st">"Time"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">time</span>, <span class="va">S</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>v <span class="op">=</span> <span class="fl">1000</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></code></pre></div>
<p><img src="https://github.com/OJWatson/magenta/raw/master/vignettes/Introduction_files/figure-html/updated%20out%20S-1.png"><!-- --></p>
<p><code>magenta</code> starts simulations close to the equilibrium solution, but it is not quite perfect. For example, in the above plot there is a burn in period before the red vertical line. The required burn in time varies depending on the simulation parameters and so it is best to run the model a few times first to work out a suitable burn in time. For most simulations though a 40 year burn in period is more than sufficient.</p>
<p>In the last simulation run, the random seed was provided, which is used to allow simulations to be able to be run reproducibly. To assist this, the function call and package version are stored with each simulation run as an attribute <code>meta</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html">attr</a></span><span class="op">(</span><span class="va">out2</span>, <span class="st">"meta"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">meta</span><span class="op">$</span><span class="va">call</span><span class="op">)</span>
<span class="co">## pipeline(EIR = 1, N = 10000, years = 10, update_length = 30, </span>
<span class="co">##     update_save = TRUE, seed = 123456789L)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">meta</span><span class="op">$</span><span class="va">version</span><span class="op">)</span>
<span class="co">## [1] '1.2.0'</span></code></pre></div>
<p><code>magenta</code> primarily characterises parasite populations in terms of a genetic barcode, in which the barcode is a vector of bits representing whether the parasite possesses the major or minor allele at the loci of interest. By default the barcode used is 24 bases long. To see what this looks like, we can request for magenta to return the human population rather than the <code>Loggers</code> as the last list element.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">out3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pipeline.html">pipeline</a></span><span class="op">(</span>EIR <span class="op">=</span> <span class="fl">1</span>, years <span class="op">=</span> <span class="fl">10</span>, N <span class="op">=</span> <span class="fl">10000</span>, 
                 update_save <span class="op">=</span> <span class="cn">TRUE</span>, update_length <span class="op">=</span> <span class="fl">30</span>, 
                 human_only_full_save <span class="op">=</span> <span class="cn">TRUE</span>, 
                 seed <span class="op">=</span> <span class="fl">123456789L</span><span class="op">)</span>
<span class="co">## Seed set to 123456789</span>
<span class="co">## magenta v1.2.0</span>
<span class="co">## Starting Stochastic Simulation for 10 years</span>
<span class="co">## Running: [=====================================================================================] 100% eta: 0s</span></code></pre></div>
<p>We can now look at what is returned in the last list element.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">out3</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">out3</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="co">## [1] "Infection_States"                             "Zetas"                                       </span>
<span class="co">## [3] "Ages"                                         "ID"                                          </span>
<span class="co">## [5] "Strain_infection_state_vectors"               "Strain_day_of_infection_state_change_vectors"</span>
<span class="co">## [7] "Strain_barcode_vectors"</span></code></pre></div>
<p>Again we return the infection states of the humans, but also their individual biting rates, <code>Zetas</code>, detection immunity and their ages, as well as the parasite strain information:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># last element</span>
<span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">out3</span><span class="op">)</span>

<span class="co"># first individual who is infected, i.e. not in states S or P (order goes S, D, A, U, T, P) </span>
<span class="va">i</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.min.html">which.min</a></span><span class="op">(</span><span class="va">out3</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">Infection_States</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span>
<span class="va">out3</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">Strain_barcode_vectors</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>
<span class="co">## [[1]]</span>
<span class="co">## [[1]][[1]]</span>
<span class="co">##  [1] FALSE  TRUE FALSE  TRUE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE  TRUE  TRUE FALSE FALSE  TRUE FALSE FALSE</span>
<span class="co">## [19] FALSE  TRUE FALSE  TRUE FALSE FALSE</span></code></pre></div>
<p>Above we have found the first individual who is infected and returned the barcodes for each parasite strain the individual has. For example, this individual had one parasite strain, as only one parasite barcode is returned. We can then look at the distribution of parasite strains in the population, for example, against their individual biting rate.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">zetas</span> <span class="op">&lt;-</span> <span class="va">out3</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">Zetas</span>
<span class="va">n_strains</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">out3</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">Strain_infection_state_vectors</span>,<span class="va">length</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">zetas</span>, <span class="va">n_strains</span>, log <span class="op">=</span> <span class="st">"x"</span>, ylab <span class="op">=</span> <span class="st">"Number of Strains"</span><span class="op">)</span></code></pre></div>
<p><img src="https://github.com/OJWatson/magenta/raw/master/vignettes/Introduction_files/figure-html/all%20infected%20individual-1.png"><!-- --></p>
<p>With the parasite barcode information we can work out what an individual’s complexity of infection (COI) is, by looking at the number of unique parasite barcodes within an individual. For example, for the individual with the most amoutn of strains:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.min.html">which.max</a></span><span class="op">(</span><span class="va">n_strains</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html">message</a></span><span class="op">(</span><span class="st">"Individual "</span>, <span class="va">m</span>, <span class="st">" has "</span>, <span class="va">n_strains</span><span class="op">[</span><span class="va">m</span><span class="op">]</span>, <span class="st">" strains"</span><span class="op">)</span>
<span class="co">## Individual 782 has 19 strains</span>
<span class="va">un</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">out3</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">Strain_barcode_vectors</span><span class="op">[[</span><span class="va">m</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html">message</a></span><span class="op">(</span><span class="st">"Individual "</span>, <span class="va">m</span>, <span class="st">" has "</span>, <span class="va">un</span>, <span class="st">" unique strains"</span><span class="op">)</span>
<span class="co">## Individual 782 has 17 unique strains</span></code></pre></div>
<p>The reason we also return the age and the detection immunity, <code>ID</code>, is that these are useful for when we want to estimate the COI in terms of what strains are likely to be observed based on their parasitaemia, which is inferred by their detection immunity, <code>ID</code>. These can be estimated using <code>convert_barcode_vectors</code> as follows, which below is used to show the COI of parsaites that have sufficiently high parasite density to be detected by microscopy if that was the only parasite strain in the infection.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cois</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/convert_barcode_vectors.html">convert_barcode_vectors</a></span><span class="op">(</span><span class="va">out3</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span>, 
                                <span class="va">out3</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">ID</span>, 
                                sub_patents_included <span class="op">=</span> <span class="cn">FALSE</span>, 
                                COI_type <span class="op">=</span> <span class="st">"patent"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">n_strains</span>, <span class="va">cois</span><span class="op">$</span><span class="va">COI</span>, ylab <span class="op">=</span> <span class="st">"Microscopy detectable COI"</span>, 
     xlab <span class="op">=</span> <span class="st">"Number of Strains"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p><img src="https://github.com/OJWatson/magenta/raw/master/vignettes/Introduction_files/figure-html/convert%20barcode%20vectors-1.png"><!-- --></p>
<p>We can also request magenta to return all the information about the human and mosquito population:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">out4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pipeline.html">pipeline</a></span><span class="op">(</span>EIR <span class="op">=</span> <span class="fl">1</span>, years <span class="op">=</span> <span class="fl">10</span>, N <span class="op">=</span> <span class="fl">10000</span>, 
                 update_save <span class="op">=</span> <span class="cn">TRUE</span>, update_length <span class="op">=</span> <span class="fl">30</span>, 
                 full_save <span class="op">=</span> <span class="cn">TRUE</span>, 
                 seed <span class="op">=</span> <span class="fl">123456789L</span><span class="op">)</span>
<span class="co">## Seed set to 123456789</span>
<span class="co">## magenta v1.2.0</span>
<span class="co">## Starting Stochastic Simulation for 10 years</span>
<span class="co">## Running: [=====================================================================================] 100% eta: 0s</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">out4</span><span class="op">[[</span><span class="va">n</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="co">## [1] "population_List"                    "populations_event_and_strains_List"</span>
<span class="co">## [3] "scourge_List"                       "parameters_List"</span></code></pre></div>
<p>The last list element now contains 4 lists:</p>
<ol style="list-style-type: decimal">
<li>
<code>population_List</code> - Human population demographic information</li>
<li>
<code>populations_event_and_strains_List</code> - Human population’s event triggers (e.g. time of next infection event, time of moving state etc) and parasite strains</li>
<li>
<code>scourge_List</code> - The mosquito population and their parasite strains and event triggers. (a scourge is the collective noun for mosquitoes)</li>
<li>
<code>parameters_List</code> - List of simulation parameters</li>
</ol>
<p>The above 4 lists are sufficient for restarting a simulation and can thus be used to save the model state to file and then be reloaded. Again see the (See <a href="https://ojwatson.github.io/magenta/articles/loggers.html">Loggers vignette</a> for more info about what information these lists actually contain, and also see the (See <a href="https://ojwatson.github.io/magenta/articles/saving.html">Saving &amp; Restarting vignette</a> for more info about how to save these lists and reload them to continue a model run.</p>
</div>
<div id="saving-different-model-summaries" class="section level3">
<h3 class="hasAnchor">
<a href="#saving-different-model-summaries" class="anchor"></a>Saving different model summaries</h3>
<p><code>magenta</code> was initially designed for simulating neutral variation and looking at patterns in COI and parasite diversity. As a result, there are a number of options for saving summaries of the within-host parasite genetic diversity at each update length.</p>
<p>Previously, we were only saving the human population’s parasites in the n-1 element. We can instead save the human population’s parasites at each update with <code>human_update_save</code>:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">out5</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pipeline.html">pipeline</a></span><span class="op">(</span>EIR <span class="op">=</span> <span class="fl">1</span>, years <span class="op">=</span> <span class="fl">10</span>, N <span class="op">=</span> <span class="fl">10000</span>, 
                 update_save <span class="op">=</span> <span class="cn">TRUE</span>, update_length <span class="op">=</span> <span class="fl">30</span>, 
                 human_update_save <span class="op">=</span> <span class="cn">TRUE</span>, 
                 summary_saves_only <span class="op">=</span> <span class="cn">FALSE</span>,
                 full_save <span class="op">=</span> <span class="cn">TRUE</span>, 
                 seed <span class="op">=</span> <span class="fl">123456789L</span><span class="op">)</span>
<span class="co">## Seed set to 123456789</span>
<span class="co">## magenta v1.2.0</span>
<span class="co">## Starting Stochastic Simulation for 10 years</span>
<span class="co">## Running: [=====================================================================================] 100% eta: 0s</span></code></pre></div>
<p>From this we could then calculate the population COI at each monthly interval amongst other genetic traits. However, saving all this information to file quickly takes up a lot of memory. Instead, we can request for the information to be summarised using <code>summary_saves_only</code>:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">out6</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pipeline.html">pipeline</a></span><span class="op">(</span>EIR <span class="op">=</span> <span class="fl">1</span>, years <span class="op">=</span> <span class="fl">10</span>, N <span class="op">=</span> <span class="fl">10000</span>, 
                 update_save <span class="op">=</span> <span class="cn">TRUE</span>, update_length <span class="op">=</span> <span class="fl">30</span>, 
                 human_update_save <span class="op">=</span> <span class="cn">TRUE</span>, 
                 summary_saves_only <span class="op">=</span> <span class="cn">TRUE</span>,
                 full_save <span class="op">=</span> <span class="cn">TRUE</span>, 
                 seed <span class="op">=</span> <span class="fl">123456789L</span><span class="op">)</span>
<span class="co">## Seed set to 123456789</span>
<span class="co">## magenta v1.2.0</span>
<span class="co">## Starting Stochastic Simulation for 10 years</span>
<span class="co">## Running: [=====================================================================================] 100% eta: 0s</span></code></pre></div>
<p>When saving summaries of the genetic information, the mean COI, % polygenomic infections, % unique barcodes and coefficient of uniqueness (see manuscript for definition) are recorded for seperate age and infection status groups. In addition, the overall clonality is recorded (how many of each barcode were there) and the population level allele frequency are saved. Lastly, some extra summaries of the population are saved, which include treatments given in the preceeding month, EIR and prevalence. For full explanation see <a href="https://ojwatson.github.io/magenta/articles/loggers.html">Loggers vignette</a>. For example, for COI the following data is stored:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">out6</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">summary_coi</span>
<span class="co">##       age_bin     clinical    N     mean</span>
<span class="co">## 1  (-0.001,5] Asymptomatic  139 3.820144</span>
<span class="co">## 2  (-0.001,5]     Clinical    2 1.000000</span>
<span class="co">## 3  (-0.001,5]            P    5 0.000000</span>
<span class="co">## 4  (-0.001,5]            S 2084 0.000000</span>
<span class="co">## 5      (5,15] Asymptomatic  413 2.987893</span>
<span class="co">## 6      (5,15]     Clinical   11 1.909091</span>
<span class="co">## 7      (5,15]            P   12 0.000000</span>
<span class="co">## 8      (5,15]            S 2472 0.000000</span>
<span class="co">## 9    (15,100] Asymptomatic  898 2.876392</span>
<span class="co">## 10   (15,100]     Clinical   14 1.285714</span>
<span class="co">## 11   (15,100]            P   23 0.000000</span>
<span class="co">## 12   (15,100]            S 3927 0.000000</span></code></pre></div>
<p>We can then plot the population mean COI over time as follows, using the last 100 values after an approximate burnin period:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># We don't want the last element as the last save</span>
<span class="va">pos</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">out6</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span>
<span class="va">COI</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">out6</span><span class="op">[</span><span class="va">pos</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span> 
    <span class="va">inf</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">$</span><span class="va">summary_coi</span><span class="op">$</span><span class="va">clinical</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Asymptomatic"</span>,<span class="st">"Clinical"</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/stats/weighted.mean.html">weighted.mean</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">summary_coi</span><span class="op">$</span><span class="va">mean</span><span class="op">[</span><span class="va">inf</span><span class="op">]</span>, <span class="va">x</span><span class="op">$</span><span class="va">summary_coi</span><span class="op">$</span><span class="va">N</span><span class="op">[</span><span class="va">inf</span><span class="op">]</span><span class="op">)</span> 
<span class="op">}</span><span class="op">)</span><span class="op">)</span>
<span class="va">prev</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">out6</span><span class="op">[</span><span class="va">pos</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="va">x</span><span class="op">$</span><span class="va">pcr_prev</span><span class="op">}</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">)</span>, mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">4</span>,<span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">tail</a></span><span class="op">(</span><span class="va">COI</span>,<span class="fl">100</span><span class="op">)</span>, xlab <span class="op">=</span> <span class="st">""</span>, ylab <span class="op">=</span> <span class="st">"COI"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">tail</a></span><span class="op">(</span><span class="va">prev</span>,<span class="fl">100</span><span class="op">)</span>, xlab <span class="op">=</span> <span class="st">""</span>, ylab <span class="op">=</span> <span class="st">"PCR PfPR"</span><span class="op">)</span></code></pre></div>
<p><img src="https://github.com/OJWatson/magenta/raw/master/vignettes/Introduction_files/figure-html/plot%20coi%20over%20time-1.png"><!-- --></p>
<p>We can increase the length of the barcode or change the assumed population level allele frequencey as well:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">out7</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pipeline.html">pipeline</a></span><span class="op">(</span>EIR <span class="op">=</span> <span class="fl">1</span>, years <span class="op">=</span> <span class="fl">10</span>, N <span class="op">=</span> <span class="fl">10000</span>, 
                 num_loci <span class="op">=</span> <span class="fl">64</span>, 
                 plaf <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">32</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0.9</span>, <span class="fl">32</span><span class="op">)</span><span class="op">)</span>,
                 update_save <span class="op">=</span> <span class="cn">TRUE</span>, 
                 update_length <span class="op">=</span> <span class="fl">30</span>, 
                 human_update_save <span class="op">=</span> <span class="cn">TRUE</span>, 
                 summary_saves_only <span class="op">=</span> <span class="cn">TRUE</span>,
                 full_save <span class="op">=</span> <span class="cn">TRUE</span>, 
                 seed <span class="op">=</span> <span class="fl">123456789L</span><span class="op">)</span>
<span class="co">## Seed set to 123456789</span>
<span class="co">## magenta v1.2.0</span>
<span class="co">## Starting Stochastic Simulation for 10 years</span>
<span class="co">## Running: [=====================================================================================] 100% eta: 0s</span></code></pre></div>
<p>And we can look at these over time:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plafs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="va">rbind</span>,<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">out7</span><span class="op">[</span><span class="va">pos</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="va">x</span><span class="op">$</span><span class="va">barcode_freq</span><span class="op">}</span><span class="op">)</span><span class="op">)</span>
<span class="fu">lattice</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/lattice/man/levelplot.html">levelplot</a></span><span class="op">(</span><span class="va">plafs</span>, xlab <span class="op">=</span> <span class="st">"Time"</span>, ylab <span class="op">=</span> <span class="st">"loci"</span><span class="op">)</span></code></pre></div>
<p><img src="https://github.com/OJWatson/magenta/raw/master/vignettes/Introduction_files/figure-html/vals%20over%20time-1.png"><!-- --></p>
<p>We can also change how we summarise the simulation at each interval. This can be done in a couple of ways. The first is by altering the age and states:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">out8</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pipeline.html">pipeline</a></span><span class="op">(</span>EIR <span class="op">=</span> <span class="fl">1</span>, years <span class="op">=</span> <span class="fl">10</span>, N <span class="op">=</span> <span class="fl">10000</span>, 
                 age_breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.1</span>,<span class="fl">5</span>,<span class="fl">10</span>,<span class="fl">15</span>,<span class="fl">20</span>,<span class="fl">100</span><span class="op">)</span>, 
                 sample_states <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span>,
                 update_save <span class="op">=</span> <span class="cn">TRUE</span>, 
                 update_length <span class="op">=</span> <span class="fl">30</span>, 
                 human_update_save <span class="op">=</span> <span class="cn">TRUE</span>, 
                 summary_saves_only <span class="op">=</span> <span class="cn">TRUE</span>,
                 full_save <span class="op">=</span> <span class="cn">TRUE</span>, 
                 seed <span class="op">=</span> <span class="fl">123456789L</span><span class="op">)</span>
<span class="co">## Seed set to 123456789</span>
<span class="co">## magenta v1.2.0</span>
<span class="co">## Starting Stochastic Simulation for 10 years</span>
<span class="co">## Running: [=====================================================================================] 100% eta: 0s</span>
<span class="va">out8</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">summary_coi</span>
<span class="co">##      age_bin     clinical   N     mean</span>
<span class="co">## 1 (-0.001,5] Asymptomatic 139 3.820144</span>
<span class="co">## 2 (-0.001,5]     Clinical   2 1.000000</span>
<span class="co">## 3     (5,15] Asymptomatic 413 2.987893</span>
<span class="co">## 4     (5,15]     Clinical  11 1.909091</span>
<span class="co">## 5   (15,100] Asymptomatic 898 2.876392</span>
<span class="co">## 6   (15,100]     Clinical  14 1.285714</span></code></pre></div>
<p>The next is by providing your own update function as <code>update_save_func</code>. Have a look at the code for <code>update_saves</code> to see the style required.</p>
</div>
<div id="interventions" class="section level3">
<h3 class="hasAnchor">
<a href="#interventions" class="anchor"></a>Interventions</h3>
<p><code>magenta</code> also allows you to specify interventions in the form of the altering the use of insecticide treated nets (<code>itn_cov</code>), indoor residual spraying (<code>irs_cov</code>) and changing the frequency of treatment (<code>ft</code>). These should be provided as vector that dictate the annual coverage.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">out9</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pipeline.html">pipeline</a></span><span class="op">(</span>EIR <span class="op">=</span> <span class="fl">4</span>, years <span class="op">=</span> <span class="fl">20</span>, N <span class="op">=</span> <span class="fl">10000</span>,
                 ft <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">10</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0.2</span>,<span class="fl">10</span><span class="op">)</span><span class="op">)</span>,
                 irs_cov <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">10</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0.2</span>,<span class="fl">10</span><span class="op">)</span><span class="op">)</span>,
                 itn_cov <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">10</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0.2</span>,<span class="fl">10</span><span class="op">)</span><span class="op">)</span>,
                 update_save <span class="op">=</span> <span class="cn">TRUE</span>, 
                 update_length <span class="op">=</span> <span class="fl">30</span>, 
                 seed <span class="op">=</span> <span class="fl">123456789L</span><span class="op">)</span>
<span class="co">## Seed set to 123456789</span>
<span class="co">## magenta v1.2.0</span>
<span class="co">## </span>
<span class="co">## Running Deterministic Model for 20 years</span>
<span class="co">## Starting Stochastic Simulation for 20 years</span>
<span class="co">## Running: [=====================================================================================] 100% eta: 0s</span>
<span class="va">infected_mosquitoes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">out9</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="va">x</span><span class="op">$</span><span class="va">Mosquitoes</span><span class="op">$</span><span class="va">Mos_I</span><span class="op">}</span><span class="op">)</span><span class="op">)</span>
<span class="va">time</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">out9</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="va">x</span><span class="op">$</span><span class="va">Time</span><span class="op">}</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fl">365</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">time</span>, <span class="va">infected_mosquitoes</span>, xlab <span class="op">=</span> <span class="st">"Time (Years)"</span>, ylab <span class="op">=</span> <span class="st">"Infected Mosquitoes"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>v <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></code></pre></div>
<p><img src="https://github.com/OJWatson/magenta/raw/master/vignettes/Introduction_files/figure-html/infected%20mosquites-1.png"><!-- --></p>
<p><code>magenta</code> incorporates vector based interventions by first running an equivalent deterministic version of the malaria transmission model with vector based intervetnions in. The output of the determininistic model is used to estimate the changing mosquito mortality rate and anthrophagy at each day as a result of the intervention changes. These are then fed through to the indivdual-based simulation to adjust the behaviour of the individual mosquitoes.</p>
</div>
<div id="space" class="section level3">
<h3 class="hasAnchor">
<a href="#space" class="anchor"></a>Space</h3>
<p><code>magenta</code> can currently be run under 2 different spatial models. The first is a closed population, with no importation. In this scenario, there will eventually be one parsaite genetic barcode that becomes fixed. To demonstrate this, we can simulate a smaller population of 1000 indiivduals, with only 4 loci in the barcode, for 25 years. The <code>spatial_type</code> is set to <code>NULL</code> here, which assumes a closed population and is also the default space option.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">out10</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pipeline.html">pipeline</a></span><span class="op">(</span>EIR <span class="op">=</span> <span class="fl">1</span>, 
                 years <span class="op">=</span> <span class="fl">25</span>, N <span class="op">=</span> <span class="fl">1000</span>, 
                 spatial_type <span class="op">=</span> <span class="cn">NULL</span>,
                 num_loci <span class="op">=</span> <span class="fl">4</span>, 
                 update_save <span class="op">=</span> <span class="cn">TRUE</span>, 
                 update_length <span class="op">=</span> <span class="fl">30</span>, 
                 human_update_save <span class="op">=</span> <span class="cn">TRUE</span>, 
                 summary_saves_only <span class="op">=</span> <span class="cn">TRUE</span>,
                 full_save <span class="op">=</span> <span class="cn">TRUE</span>, 
                 seed <span class="op">=</span> <span class="fl">123456789L</span><span class="op">)</span>
<span class="co">## Seed set to 123456789</span>
<span class="co">## magenta v1.2.0</span>
<span class="co">## Starting Stochastic Simulation for 25 years</span>
<span class="co">## Running: [=====================================================================================] 100% eta: 0s</span></code></pre></div>
<p>We can see this process again by plotting the barcode frequency over time for the fist n-1 elements (as the last list element will not contain the barcode_frequency summary)</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pos</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">out10</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span>
<span class="va">plafs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="va">rbind</span>,<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">out10</span><span class="op">[</span><span class="va">pos</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="va">x</span><span class="op">$</span><span class="va">barcode_freq</span><span class="op">}</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/matplot.html">matplot</a></span><span class="op">(</span><span class="va">plafs</span><span class="op">)</span></code></pre></div>
<p><img src="https://github.com/OJWatson/magenta/raw/master/vignettes/Introduction_files/figure-html/fixation%20lattice-1.png"><!-- --></p>
<p>We can change this to have importation of barcodes over time. This occurs with a genetic island model where the population we are simulating is assumed to receive imported parasites from a genetically stable population. There are two forms of importation, the <code>spatial_incidence_matrix</code> and <code>spatial_mosquitoFOI_matrix</code>, which determine the rate at which new parasites are imported into the human and mosquito population respectively. In the following we assume that 20% of new infections in both the humans and mosquito populations are actually imported, with the barcodes drawn according to the population level allele frequency (PLAF) (by default the PLAF is 0.5 for each loci). This ensures that a genetic equilibrium is reached and barcode fixation does not occur.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">out11</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pipeline.html">pipeline</a></span><span class="op">(</span>EIR <span class="op">=</span> <span class="fl">1</span>, 
                 years <span class="op">=</span> <span class="fl">25</span>, 
                 N <span class="op">=</span> <span class="fl">1000</span>, 
                 spatial_type <span class="op">=</span> <span class="st">"island"</span>,
                 spatial_incidence_matrix <span class="op">=</span> <span class="fl">0.2</span>,
                 spatial_mosquitoFOI_matrix <span class="op">=</span> <span class="fl">0.2</span>,
                 num_loci <span class="op">=</span> <span class="fl">4</span>, 
                 update_save <span class="op">=</span> <span class="cn">TRUE</span>, 
                 update_length <span class="op">=</span> <span class="fl">30</span>, 
                 human_update_save <span class="op">=</span> <span class="cn">TRUE</span>, 
                 summary_saves_only <span class="op">=</span> <span class="cn">TRUE</span>,
                 full_save <span class="op">=</span> <span class="cn">TRUE</span>, 
                 seed <span class="op">=</span> <span class="fl">123456789L</span><span class="op">)</span>
<span class="co">## Seed set to 123456789</span>
<span class="co">## magenta v1.2.0</span>
<span class="co">## Starting Stochastic Simulation for 25 years</span>
<span class="co">## Running: [=====================================================================================] 100% eta: 0s</span>
<span class="va">pos</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">out11</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span>
<span class="va">plafs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="va">rbind</span>,<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">out11</span><span class="op">[</span><span class="va">pos</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="va">x</span><span class="op">$</span><span class="va">barcode_freq</span><span class="op">}</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/matplot.html">matplot</a></span><span class="op">(</span><span class="va">plafs</span><span class="op">)</span></code></pre></div>
<p><img src="https://github.com/OJWatson/magenta/raw/master/vignettes/Introduction_files/figure-html/importation-1.png"><!-- --></p>
</div>
<div id="seasonality-and-specific-locations" class="section level3">
<h3 class="hasAnchor">
<a href="#seasonality-and-specific-locations" class="anchor"></a>Seasonality and Specific Locations</h3>
<p>Included in <code>magenta</code> are the historic interventions and seasonality profiles for admin level one regions in sub-Saharan African countries. These can then be used to draw up the required parameter sets for a given region. E.g. to simulate for Kinsasha from 1980:2015</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># set up the years</span>
<span class="va">year_range</span> <span class="op">&lt;-</span> <span class="fl">1980</span><span class="op">:</span><span class="fl">2015</span>

<span class="co"># grab the interventions for region of interest</span>
<span class="va">country</span> <span class="op">&lt;-</span> <span class="st">"Democratic Republic of the Congo"</span>
<span class="va">admin</span> <span class="op">&lt;-</span> <span class="st">"Kinshasa"</span>
<span class="va">ints</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/intervention_grab.html">intervention_grab</a></span><span class="op">(</span>country <span class="op">=</span> <span class="va">country</span>, 
                          admin <span class="op">=</span> <span class="va">admin</span>,
                          year_range <span class="op">=</span> <span class="va">year_range</span><span class="op">)</span>
<span class="co">## Requested: Kinshasa, Democratic Republic of the Congo</span>
<span class="co">## Returned: Kinshasa, Democratic Republic of the Congo</span>
<span class="co"># grab the importation from the migration model</span>
<span class="va">spl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spl_grab.html">spl_grab</a></span><span class="op">(</span><span class="va">country</span>, <span class="va">admin</span>, <span class="va">year_range</span><span class="op">)</span>
<span class="co">## Requested: Kinshasa, Democratic Republic of the Congo</span>
<span class="co">## Returned: Kinshasa_City, Democratic Republic of the Congo</span>
<span class="co"># get the intervention parmeters</span>
<span class="va">ft</span> <span class="op">&lt;-</span> <span class="va">ints</span><span class="op">$</span><span class="va">ft</span>
<span class="va">itn_cov</span> <span class="op">&lt;-</span> <span class="va">ints</span><span class="op">$</span><span class="va">itn_cov</span>
<span class="va">irs_cov</span> <span class="op">&lt;-</span> <span class="va">ints</span><span class="op">$</span><span class="va">irs_cov</span>

<span class="co"># load the Malaria Atlas Project Prevalence data</span>
<span class="va">MAP_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/MAP_PfPR_population_weighted.csv"</span>, 
                        package <span class="op">=</span> <span class="st">"magenta"</span><span class="op">)</span>
<span class="va">MAP</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span><span class="va">MAP_file</span><span class="op">)</span>

<span class="co"># work out the EIR for the prevalence in year 2000 (i.e. pre interventions)</span>
<span class="va">pfpr_micro</span> <span class="op">&lt;-</span> <span class="va">MAP</span><span class="op">$</span><span class="va">`2000`</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">MAP</span><span class="op">$</span><span class="va">Name</span><span class="op">==</span><span class="va">admin</span><span class="op">)</span><span class="op">]</span>
<span class="va">EIR</span> <span class="op">&lt;-</span> <span class="fu">magenta</span><span class="fu">:::</span><span class="fu"><a href="../reference/pfpr_to_eir_heuristic.html">pfpr_to_eir_heuristic</a></span><span class="op">(</span>ft <span class="op">=</span> <span class="fl">0</span>, PfPR_micro <span class="op">=</span> <span class="va">pfpr_micro</span><span class="op">)</span>
<span class="co">## 0.00152858753907675</span>
<span class="co"># run the simulation</span>
<span class="va">out12</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pipeline.html">pipeline</a></span><span class="op">(</span>
  EIR <span class="op">=</span> <span class="va">EIR</span>,
  N <span class="op">=</span> <span class="fl">10000</span>,
  years <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">itn_cov</span><span class="op">)</span>,
  country <span class="op">=</span> <span class="va">country</span>,
  admin <span class="op">=</span> <span class="va">admin</span>,
  itn_cov <span class="op">=</span> <span class="va">itn_cov</span>,
  irs_cov <span class="op">=</span> <span class="va">irs_cov</span>,
  ft <span class="op">=</span> <span class="va">ft</span>,
  spatial_incidence_matrix <span class="op">=</span> <span class="va">spl</span><span class="op">$</span><span class="va">incidence</span>,
  spatial_mosquitoFOI_matrix <span class="op">=</span> <span class="va">spl</span><span class="op">$</span><span class="va">mosquitoFOI</span>,
  spatial_type <span class="op">=</span> <span class="st">"island"</span>,
  human_only_full_save <span class="op">=</span> <span class="cn">TRUE</span>,
  update_length <span class="op">=</span> <span class="fl">30</span>,
  update_save <span class="op">=</span> <span class="cn">TRUE</span>,
  human_update_save <span class="op">=</span> <span class="cn">FALSE</span>,
  summary_saves_only <span class="op">=</span> <span class="cn">FALSE</span>,
<span class="op">)</span>
<span class="co">## Seed set to 693175740</span>
<span class="co">## magenta v1.2.0</span>
<span class="co">## Requested: Kinshasa, Democratic Republic of the Congo</span>
<span class="co">## Returned: Kinshasa, Democratic Republic of the Congo</span>
<span class="co">## </span>
<span class="co">## Running Deterministic Model for 36 years</span>
<span class="co">## Starting Stochastic Simulation for 36 years</span>
<span class="co">## Running: [=====================================================================================] 100% eta: 0s</span>
<span class="co"># Grab the microscopy prevalence over time in 2-10s</span>
<span class="va">pos</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">out12</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span>
<span class="va">time</span> <span class="op">&lt;-</span> <span class="fl">1980</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">out12</span><span class="op">[</span><span class="va">pos</span><span class="op">]</span>, <span class="st">"[["</span>, <span class="st">"Time"</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fl">365</span>
<span class="va">prev_2_10_micro</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">out12</span><span class="op">[</span><span class="va">pos</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span> 
  <span class="fu">magenta</span><span class="fu">:::</span><span class="fu">microscopy_detected</span><span class="op">(</span>ages <span class="op">=</span> <span class="va">x</span><span class="op">$</span><span class="va">Ages</span>, 
                                IDs <span class="op">=</span> <span class="va">x</span><span class="op">$</span><span class="va">ID</span>, 
                                states <span class="op">=</span> <span class="va">x</span><span class="op">$</span><span class="va">InfectionStates</span><span class="op">)</span>
  <span class="op">}</span><span class="op">)</span><span class="op">)</span>

<span class="co"># get the MAP estimates</span>
<span class="va">map_prev</span> <span class="op">&lt;-</span> <span class="va">MAP</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">MAP</span><span class="op">$</span><span class="va">Name</span><span class="op">==</span><span class="va">admin</span><span class="op">)</span>,<span class="fl">5</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span>

<span class="co"># plot the relationship</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">time</span>, <span class="va">prev_2_10_micro</span>, xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2000</span>,<span class="fl">2015</span><span class="op">)</span>, 
     ylab <span class="op">=</span> <span class="st">"Micro 2-10"</span>, type <span class="op">=</span> <span class="st">"l"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/points.html">points</a></span><span class="op">(</span><span class="fl">2000</span><span class="op">:</span><span class="fl">2015</span>, <span class="va">map_prev</span><span class="op">)</span></code></pre></div>
<p><img src="https://github.com/OJWatson/magenta/raw/master/vignettes/Introduction_files/figure-html/country%20example-1.png"><!-- --></p>
<p>The above is similar to how the model fitting simulations were laid out and it gives the user a chance to check that the parameter sets being used are the correct parameters. However, you can use <code>use_historic_interventions</code> as <code>TRUE</code> and provide the year range and <code>pipeline</code> will conduct the above parameter searching for you.</p>
<p>As an aside, <code>magenta</code> has a multitude of little helped functions like <code>microscopy_detected</code> that are not exported currently, due to their use in mainly assisting downstream analysis rather than being for simulation directly.</p>
</div>
<div id="ibd-simulations" class="section level3">
<h3 class="hasAnchor">
<a href="#ibd-simulations" class="anchor"></a>IBD Simulations</h3>
<p><code>magenta</code> was also designed to simulate patterns in identity-by-descent (IBD). In these simulations, the barcode is adapted such that each loci is represented by multiple bits that represent a unique integer rather than a single bit. The integers are assigned at the beginning of the simulation, with consecutive integers assigned to each parasite, which allows the pedigree of parasites to be tracked. These simulations are slower as the bitset container used to represnt is substantially larger and simulating recombination and new variants is slightly more involved. In addition, summarising the parasite population at monthly intervals takes more time and more memory is required to simulate patterns of IBD.</p>
<p>To simulate IBD, we allocate the <code>ibd_length</code>, which is the bumber of bits used to represent one locus. The number of bits used gives the maximum number of parasite pedigrees that can be tracked, which is equal to <code>2^ibd_length</code>. As a result you need to double check that you will not exceed this number during a simulation. An easy way to check is to use the argument <code>set_up_only</code>, which will return the simulation model state after it has been initialised:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">out13</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pipeline.html">pipeline</a></span><span class="op">(</span>EIR <span class="op">=</span> <span class="fl">1</span>, 
                  years <span class="op">=</span> <span class="fl">10</span>, 
                  N <span class="op">=</span> <span class="fl">1000</span>,  
                  ibd_length <span class="op">=</span> <span class="fl">20</span>, 
                  num_loci <span class="op">=</span> <span class="fl">24</span>,
                  set_up_only <span class="op">=</span> <span class="cn">TRUE</span>,
                  update_save <span class="op">=</span> <span class="cn">TRUE</span>, 
                  update_length <span class="op">=</span> <span class="fl">30</span>, 
                  human_update_save <span class="op">=</span> <span class="cn">TRUE</span>, 
                  summary_saves_only <span class="op">=</span> <span class="cn">TRUE</span>,
                  full_save <span class="op">=</span> <span class="cn">TRUE</span>, 
                  seed <span class="op">=</span> <span class="fl">123456789L</span><span class="op">)</span>
<span class="co">## Seed set to 123456789</span>
<span class="co">## magenta v1.2.0</span>
<span class="co">## Set Up Grab</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">out13</span><span class="op">)</span>
<span class="co">## [1] "population_List"                    "populations_event_and_strains_List"</span>
<span class="co">## [3] "scourge_List"                       "parameters_List"</span></code></pre></div>
<p>From this we can have a look at the strains to see the IBD representation used:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">infected</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">out13</span><span class="op">$</span><span class="va">population_List</span><span class="op">$</span><span class="va">Infection_States</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span><span class="op">)</span>
<span class="va">strains</span> <span class="op">&lt;-</span> <span class="va">out13</span><span class="op">$</span><span class="va">populations_event_and_strains_List</span><span class="op">$</span><span class="va">Strain_barcode_vectors</span><span class="op">[</span><span class="va">infected</span><span class="op">]</span>

<span class="va">strains</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>
<span class="co">## [[1]]</span>
<span class="co">## [[1]][[1]]</span>
<span class="co">##   [1] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">##  [19] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">##  [37] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">##  [55] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">##  [73] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">##  [91] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [109] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [127] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [145] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [163] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [181] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [199] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [217] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [235] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [253] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [271] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [289] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [307] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [325] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [343] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [361] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [379] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [397] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [415] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [433] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [451] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [469] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## </span>
<span class="co">## [[1]][[2]]</span>
<span class="co">##   [1]  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">##  [19] FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">##  [37] FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">##  [55] FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">##  [73] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">##  [91] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [109] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [127] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE</span>
<span class="co">## [145] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE</span>
<span class="co">## [163] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [181]  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [199] FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [217] FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [235] FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [253] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [271] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [289] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [307] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE</span>
<span class="co">## [325] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE</span>
<span class="co">## [343] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [361]  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [379] FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [397] FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [415] FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [433] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [451] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span>
<span class="co">## [469] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span></code></pre></div>
<p>The first infected individual is infected with two parasites as indicated with the two strain barcode vectors. The first is the binary representation of <code><a href="https://rdrr.io/r/base/rep.html">rep(0, 24)</a></code> and the next is <code><a href="https://rdrr.io/r/base/rep.html">rep(1,24)</a></code>. To convert these to integer vectors we can use the unexported <code>convert_ibd_barcode</code> passing in the barcode and the number of loci (<code>nl</code>) being simulated. So the first parasite in the second infected individual should be represented by a barcode of 2s give that the first individual had two strains (i.e. the 0s and the 1s).</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">magenta</span><span class="fu">:::</span><span class="fu">convert_ibd_barcode</span><span class="op">(</span><span class="va">strains</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, nl <span class="op">=</span> <span class="fl">24</span><span class="op">)</span>
<span class="co">##  [1] 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2</span></code></pre></div>
<p>We can then work out how many parasites have been allocated in the simulation to begin with:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">last_infected</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">tail</a></span><span class="op">(</span><span class="va">strains</span>,<span class="fl">1</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="fu">magenta</span><span class="fu">:::</span><span class="fu">convert_ibd_barcode</span><span class="op">(</span><span class="va">last_infected</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">last_infected</span><span class="op">)</span><span class="op">]</span><span class="op">]</span>, nl <span class="op">=</span> <span class="fl">24</span><span class="op">)</span>
<span class="co">##  [1] 607 607 607 607 607 607 607 607 607 607 607 607 607 607 607 607 607 607 607 607 607 607 607 607</span></code></pre></div>
<p>Clearly <code>2^20</code> will be plenty, even if we add in substantial importation, with importation in the IBD model importing the next integer barcode, i.e. <code><a href="https://rdrr.io/r/base/rep.html">rep(608, 24)</a></code>. Just remember you need to factor in both the rate of importation, the length of the simulation and the size of the population. We can do the same check at the end of the simulation to make sure we allocated sufficient bits.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">out14</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pipeline.html">pipeline</a></span><span class="op">(</span>EIR <span class="op">=</span> <span class="fl">1</span>, 
                  years <span class="op">=</span> <span class="fl">10</span>, 
                  N <span class="op">=</span> <span class="fl">1000</span>,  
                  ibd_length <span class="op">=</span> <span class="fl">20</span>, 
                  num_loci <span class="op">=</span> <span class="fl">24</span>,
                  spatial_type <span class="op">=</span> <span class="st">"island"</span>,
                  spatial_incidence_matrix <span class="op">=</span> <span class="fl">0.2</span>,
                  spatial_mosquitoFOI_matrix <span class="op">=</span> <span class="fl">0.2</span>,
                  update_save <span class="op">=</span> <span class="cn">TRUE</span>, 
                  update_length <span class="op">=</span> <span class="fl">30</span>, 
                  human_update_save <span class="op">=</span> <span class="cn">TRUE</span>, 
                  summary_saves_only <span class="op">=</span> <span class="cn">TRUE</span>,
                  full_save <span class="op">=</span> <span class="cn">TRUE</span>, 
                  seed <span class="op">=</span> <span class="fl">123456789L</span><span class="op">)</span>
<span class="co">## Seed set to 123456789</span>
<span class="co">## magenta v1.2.0</span>
<span class="co">## Starting Stochastic Simulation for 10 years</span>
<span class="co">## Running: [=====================================================================================] 100% eta: 0s</span>
<span class="va">l</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">out14</span><span class="op">)</span>
<span class="va">strains</span> <span class="op">&lt;-</span> <span class="va">out14</span><span class="op">[[</span><span class="va">l</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">populations_event_and_strains_List</span><span class="op">$</span><span class="va">Strain_barcode_vectors</span>
<span class="va">infected</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">out14</span><span class="op">[[</span><span class="va">l</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">population_List</span><span class="op">$</span><span class="va">Infection_States</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span><span class="op">)</span>

<span class="va">converted_barcodes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">strains</span><span class="op">[</span><span class="va">infected</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">x</span>, <span class="fu">magenta</span><span class="fu">:::</span><span class="va">convert_ibd_barcode</span>, <span class="fl">24</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">converted_barcodes</span><span class="op">)</span><span class="op">)</span>
<span class="co">## [1] 9025</span></code></pre></div>
<p>Definitely plenty.</p>
<p>As when simulating normal barcodes, a summary of the genetics is made at each update interval. For simulations of IBD, we save the population level IBD (pIBD in the manuscript) as <code>summary_ibd</code> and the within host IBD (iIBD in the manuscript) as <code>summary_within_ibd</code>. The same system is used for how to bin the population into clinical infection and age groups:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">out14</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">summary_ibd</span>
<span class="co">##       age_bin     clinical  N         mean</span>
<span class="co">## 1  (-0.001,5] Asymptomatic 14 9.751341e-06</span>
<span class="co">## 2  (-0.001,5]            P  0          NaN</span>
<span class="co">## 3  (-0.001,5]            S  0          NaN</span>
<span class="co">## 4      (5,15] Asymptomatic 27 1.560287e-04</span>
<span class="co">## 5      (5,15]            P  0          NaN</span>
<span class="co">## 6      (5,15]            S  0          NaN</span>
<span class="co">## 7    (15,100] Asymptomatic 59 1.875604e-04</span>
<span class="co">## 8    (15,100]     Clinical  1 3.154574e-03</span>
<span class="co">## 9    (15,100]            P  0          NaN</span>
<span class="co">## 10   (15,100]            S  0          NaN</span></code></pre></div>
<p>And for iIBD:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">out14</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">summary_within_ibd</span>
<span class="co">##       age_bin     clinical  N       mean</span>
<span class="co">## 1  (-0.001,5] Asymptomatic  9 0.02314815</span>
<span class="co">## 2  (-0.001,5]            P  0        NaN</span>
<span class="co">## 3  (-0.001,5]            S  0        NaN</span>
<span class="co">## 4      (5,15] Asymptomatic 19 0.00000000</span>
<span class="co">## 5      (5,15]            P  0        NaN</span>
<span class="co">## 6      (5,15]            S  0        NaN</span>
<span class="co">## 7    (15,100] Asymptomatic 24 0.00000000</span>
<span class="co">## 8    (15,100]     Clinical  0        NaN</span>
<span class="co">## 9    (15,100]            P  0        NaN</span>
<span class="co">## 10   (15,100]            S  0        NaN</span></code></pre></div>
<p>Note how the denominator is different between the two measures. This is because iIBD is the comparison of IBD (same integer at a given locus) of the strains within an individual. Thus if an individual is only infected with one parasite strain then iIBD can not be calculated and these individuals are not included.</p>
</div>
</div>
<div id="summary" class="section level2">
<h2 class="hasAnchor">
<a href="#summary" class="anchor"></a>Summary</h2>
<p>The above introduction gives an overview of how to run the model and covers all components of <code>magenta</code> that were used to produce the findings in the <a href="https://www.biorxiv.org/content/10.1101/793554v1">BioArxiv manuscript</a>.</p>
<p><code>magenta</code>, however, has been developed to also simulate drug resistance and consider more nuanced spatial set ups. Both of these are under development, with the resistance implementation fully in (but needing more documentation) and the inclusion of space via a metapopulation not fully implemented yet. Once documented these will be availabe in the <a href="https://ojwatson.github.io/magenta/articles/resistance.html">Resistance Vignette</a> and <a href="https://ojwatson.github.io/magenta/articles/spatial.html">Spatial Vignette</a></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by OJ Watson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
